### PR TITLE
ARROW-13482: [C++][Compute] Refactoring away from hard coded ExecNode factories to a registry

### PR DESCRIPTION
**Pull request by bkietz**
- An extensible registry of exec node factories (`std::function<Result<ExecNode*>(ExecPlan* plan, std::vector<ExecNode*> inputs, const ExecNodeOptions& options)>`) is provided
- Hard coded factories like `compute::MakeSinkNode`, `dataset::MakeScanNode` are replaced by factories in the registry named "sink", "scan", etc
- `arrow::compute::Declaration` is provided to represent an unconstructed set of `ExecNode`s, which can be validated and emplaced into an `ExecPlan` as a unit

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]
LGTM, thanks. Overall this looks cleaner than hand-constructing the graphs before, at least for the simple ones we have in tests.

**Review by nirandaperera — CHANGES_REQUESTED** [review]
Do you think we could split the `exec_plan.cc` based on each ExecNode impl? It would be much easier to navigate then

**Comment by bkietz** [comment]
Will do, thanks!

**Comment by ianmcook** [comment]
Is there any plan for documenting these ExecNode factories in a standard way? Does this registry give us opportunities to achieve ARROW-13227 in a modular way?

**Comment by bkietz** [comment]
@ianmcook I'll add a note to ARROW-13227 describing the impact of this PR. In short, ARROW-13227 should now document these factories instead of specific subclasses of ExecNode (and will look a lot like `compute.rst`).

**Review by lidavidm — APPROVED** [review]
LGTM, thanks for this (massive) refactor! 

I left one small nit.

File: cpp/src/arrow/compute/exec/aggregate_node.cc [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+    const std::vector<internal::Aggregate>& aggregates,
+    const std::vector<ValueDescr>& in_descrs);
+
+Result<FieldVector> ResolveKernels(
+    const std::vector<internal::Aggregate>& aggregates,
+    const std::vector<const HashAggregateKernel*>& kernels,
+    const std::vector<std::unique_ptr<KernelState>>& states, ExecContext* ctx,
+    const std::vector<ValueDescr>& descrs);
+
+}  // namespace internal
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/compute/exec/aggregate_node.cc:58]: I was thinking I'd replace these with usage of `Kernel::InitAll` or other public & tested utilities in https://issues.apache.org/jira/browse/ARROW-13451. I've added a note to this effect to that JIRA

File: cpp/src/arrow/compute/exec/aggregate_node.cc [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+  const std::vector<const HashAggregateKernel*> agg_kernels_;
+
+  ThreadIndexer get_thread_index_;
+  AtomicCounter input_counter_, output_counter_;
+
+  std::vector<ThreadLocalState> local_states_;
+  ExecBatch out_data_;
+};
+
+class DefaultExecFactoryRegistry : public ExecFactoryRegistry {
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/compute/exec/aggregate_node.cc:584]: whoops, didn't intend to leave this here

**Comment by bkietz** [comment]
+1, merging

**Pull request closed without merging**